lib/HLSL/DxilSimpleGVNHoist.cpp

```diff
@@ -585,3 +593,197 @@ FunctionPass *llvm::createDxilSimpleGVNHoistPass() {
+// - Find the common destination 'End' of the branches.
+// - Find Find two predecessors 'P0' and 'P1' of 'End' such that 'S0' dominates
+//   'P0' and 'P0' post-dominates 'S0', and 'P0' only has a single successor
+//   (Same with 'S1' and 'P1').
+//   This means if 'Br'->'S0' is taken, then 'End' must be reached via 'P0'; if
+//   'End' is reached via 'P0', 'Br'->'S0' must have been taken (Same with 'S1'
+//   and 'P1').
+// - Using ValueTable, compare if every pair of incoming values from P0 and P1
+//   is identical for any PHIs in 'End'
+// - Make sure there is no side effect or loop between 'Br' and 'End'
+// - If all above checks succeed, replace 'Br' with with an unconditional branch
+//   to S0
+// 
+// The current state of the pass is pretty limited. If incoming values from P0
+// and P1 are dependent on any PHIs defined between Br and End, then the pass
+// will fail to simplify the branch. If there are any side effects within the
+// region, the pass will also fail. It's possible to handling these cases, but
+// require proving the two sides of the branch have equivalent control flow,
+// which is non-trivial, and will be left to a later date.
+//
+namespace {
+
+class DxilSimpleGVNEliminateRegion : public FunctionPass {
+
+  bool CheckRegionForSideEffectsOrLoops(BasicBlock *Begin, BasicBlock *End /*Non inclusive*/);
+
+  std::unordered_map<BasicBlock *, bool> BlockHasSideEffects;
+  bool MayHaveSideEffects(BasicBlock *BB) {
+    auto It = BlockHasSideEffects.find(BB);
+    if (It != BlockHasSideEffects.end())
+      return It->second;
+
+    for (Instruction &I : *BB) {
+      if (I.mayHaveSideEffects() && !hlsl::IsNop(&I)) {
+        BlockHasSideEffects[BB] = true;
+        return true;
+      }
+    }
+
+    BlockHasSideEffects[BB] = false;
+    return false;
+  }
+  bool ProcessBB(BasicBlock &BB, ValueTable &VT, DominatorTree *DT, PostDominatorTree *PDT);
+
+public:
+  static char ID; // Pass identification, replacement for typeid
+  explicit DxilSimpleGVNEliminateRegion() : FunctionPass(ID) {}
+
+  StringRef getPassName() const override {
+    return "DXIL simple GVN eliminate region";
+  }
+
+  bool runOnFunction(Function &F) override;
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.addRequired<PostDominatorTree>();
+    AU.addRequired<DominatorTreeWrapperPass>();
+  }
+};
+
+char DxilSimpleGVNEliminateRegion::ID = 0;
+
+bool DxilSimpleGVNEliminateRegion::CheckRegionForSideEffectsOrLoops(BasicBlock *Begin, BasicBlock *End /*Non inclusive*/) {
+  SmallVector<BasicBlock *, 10> Worklist;
+  Worklist.push_back(Begin);
+  SmallPtrSet<BasicBlock *, 10> Seen;
+
+  while (Worklist.size()) {
+    BasicBlock *BB = Worklist.pop_back_val();
+    // Stop before reaching End
+    if (BB == End)
+      continue;
+
+    if (MayHaveSideEffects(BB))
+      return false;
+
+    Seen.insert(BB);
+    for (BasicBlock *Succ : successors(BB)) {
+      // Goes back into the region. Give up.
+      if (Seen.count(Succ))
```

Comment:
If the region is wholly contained in a loop you would remove it, but there is no test case for that. We should add one.

Reply:
Done.